sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala

```diff
@@ -78,15 +78,20 @@ case class TransformWithStateExec(
   override def shortName: String = "transformWithStateExec"
 
   override def shouldRunAnotherBatch(newInputWatermark: Long): Boolean = {
-    timeMode match {
-      case ProcessingTime =>
-        // TODO: check if we can return true only if actual timers are registered, or there is
-        // expired state
-        true
-      case EventTime =>
-        eventTimeWatermarkForEviction.isDefined &&
-          newInputWatermark > eventTimeWatermarkForEviction.get
-      case _ => false
+    if (outputMode == OutputMode.Append || outputMode == OutputMode.Update) {
```

Comment:
Hmm - won't we have an issue with `Append` mode using proc time based time mode ?

Reply:
This condition should always be true for timeMode set to ProcessingTime. I have refactored this condition.